core/client/hdfs/src/main/java/alluxio/hadoop/AlluxioHdfsInputStream.java

```diff
@@ -50,7 +51,20 @@ public int read() throws IOException {
 
   @Override
   public int read(ByteBuffer buf) throws IOException {
-    return mInput.read(buf);
+    if (mInput.getWrappedStream() instanceof ByteBufferReadable) {
+      return mInput.read(buf);
+    } else {
+      int off = buf.position();
+      int len = buf.remaining();
+      byte[] byteArray = new byte[len];
```

Comment:
You can save a copy if the `buf` is backed by a bytes array. You retrieve the array by `buf.hasArray()` and `buf.array()`

Reply:
Done